rasterio/rio/shapes.py

```diff
@@ -109,21 +106,30 @@ def shapes(
     if not sequence:
         geojson_type = 'collection'
 
+    geographic = True if projection == 'geographic' else False
+
     try:
-        with ctx.obj['env'] as env:
-            write_features(
-                stdout,
-                feature_gen(
-                    input, env, bidx, sampling,  band, as_mask, with_nodata, projection, precision),
-                sequence=sequence,
-                geojson_type=geojson_type, use_rs=use_rs,
-                **dump_kwds)
+        with rasterio.open(input) as src:
+            with ctx.obj['env'] as env:
```

Comment:
@perrygeo it's best to make `env` the outer context so it can affect dataset opening when needed. I think this is the last todo in the PR.